reconcile/dashdotdb_dora.py

```diff
@@ -0,0 +1,538 @@
+                        AND trigger_reason LIKE {}
+                        AND ({})
+                    ORDER BY deployment_time ASC;
+                """
+            ).format(sql.Literal(trigger_reason_pattern), sql.SQL(" OR ").join(subq))
+
+            cur.execute(query)
+
+            for record in cur:
+                deployments.append(
+                    (AppEnv(record[0], record[1]), Deployment(record[2], record[3]))
+                )
+
+        if not deployments:
+            LOG.info("No deployments found")
+
+        return deployments
+
+
+@dataclass
+class Commit:
+    repo: str
+    sha: str
+    date: datetime
+
+    def lttc(self, finish_timestamp: datetime) -> int:
+        commit_date_tzaware = self.date
+        finish_timestamp_tzaware = finish_timestamp
+
+        if commit_date_tzaware.tzinfo is None:
+            commit_date_tzaware = commit_date_tzaware.replace(tzinfo=timezone.utc)
+
+        if finish_timestamp_tzaware.tzinfo is None:
+            finish_timestamp_tzaware = finish_timestamp_tzaware.replace(
+                tzinfo=timezone.utc
+            )
+
+        return int((finish_timestamp_tzaware - commit_date_tzaware).total_seconds())
+
+
+class DashdotdbDORA(DashdotdbBase):
+    """DashdotDB DORA collector.
+
+    Definitions:
+
+    * `app_name` is the value of `.name` in the saas-file.
+    * `env_name` is the value of a given
+      `.resourceTemplates.targets.namespace.$ref -> ns file -> environment file
+      -> .name`.
+    * `trigger_reason` is one of the fields registered in DeploymentDB that
+      identifies the promotion event. It looks like:
+      https://gitlab.../service/app-interface/commit/<sha>.
+    * `promotion_sha` is the sha contained in the trigger_reason.
+
+    This collector performs the followig tasks:
+
+    * Get a list of saasfiles that contain the dora field in its labels. The
+      value is a comma-separated list of target env_names. Relevant MR.
+    * Query DashdotDB (over API) to obtain latest registered entry for that
+      app_name and env_name pair.
+    * Query DeploymentDB (over SQL) to obtain the deployments associated to each
+      app_name and env_name pair from the latest registered entry in DashdotDB
+      or from the last 90 days if there are none.
+    * From the DeploymentDB entry trigger_reason, extract the changes: url of
+      the repository, previous commit, promotion commit. This is done by
+      fetching the saas-file from GitLab's API at at the promotion_sha commit,
+      and comparing it with the promotion_sha^ commit (note the ^).
+    * Using GitHub or GitLab's API, obtain the list of commits between
+      promotion_sha^ and promotion_sha.
+    * Build the payload for DashdotDB and submit it as a POST.
+
+    Caveats:
+
+    * This class has been designed to run as a cronjob, so the cleanup code
+      should be reviewed if this changes its execution pattern to a service.
+    """
+
+    def __init__(
```

Comment:
Now I know why you must mock your `DashdotdbDORA.__init__` in your test cases. Using the dependency injection pattern to decouple your class for easier testing would be best.

```python
    def __init__(
        self,
        dry_run: bool,
        gitlab_api: GitLabApi,
        github_token: str,
        secret_reader: SecretReaderBase,
        thread_pool_size: int = 5,
    ) -> None:
```

The best is you don't need to query app-interface in your `__init__` method.

Reply:
Agreed, but going to leave this for the following iteration.